File: Leetcode/sort_and_search/367_Valid Perfect Square/Solution.py
class Solution(object):
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        """
        if num==1: return True
        start, end = 0, num//2
        while start <= end:
            mid = (start+end)//2
            if mid**2==num:
                return True
            if mid**2 > num:
                end = mid - 1
            else:
                start = mid + 1
        return False

File: Leetcode/sort_and_search/367_Valid Perfect Square/Solution.java
class Solution {
    public boolean isPerfectSquare(int num) {
        if(num==1) return true;
        long start = 0, end = num/2;
        while(start<=end){
            long mid = (start+end)/2;
            if(mid*mid == (long)num) return true;
            if( mid*mid>(long)(num)){
                end = mid - 1;
            }else{
                start = mid + 1;
            }
        }
        return false;
    }
}
